framework/src/base/Assembly.C

```diff
@@ -3324,21 +3322,33 @@ Assembly::addJacobianBlock(SparseMatrix<Number> & jacobian,
                            const DofMap & dof_map,
                            std::vector<dof_id_type> & dof_indices)
 {
+  if (dof_indices.size() == 0)
+    return;
+  if (!(*_cm)(ivar, jvar))
+    return;
+
   DenseMatrix<Number> & ke = jacobianBlock(ivar, jvar);
+  auto & iv = _sys.getVariable(_tid, ivar);
+  auto & jv = _sys.getVariable(_tid, jvar);
+  auto & scaling_factor = iv.arrayScalingFactor();
 
-  // stick it into the matrix
-  std::vector<dof_id_type> di(dof_indices);
-  dof_map.constrain_element_matrix(ke, di, false);
+  unsigned int i = ivar - iv.number();
```

Comment:
It is guaranteed by design `iv.number <= ivar` since `iv` is obtained through `SystemBase::getVariable` with `ivar`. Most of times `ivar` will just be equal to `iv.number` except for array variables, where ivar could be a number for a component of an array variable but calling `getVariable` will return the array variable that has the number of the 0th component.

Reply:
Ok, I think you should put an in-line comment here then explaining what you just explained to me